### PR TITLE
Fixes major issues and (appears) to work properly

### PR DESCRIPTION
**Pull request by toroidal-code**
I had trouble building, and fixed up the glaring issues. There's also a Makefile, since I thought that might be a good idea.

I'm using libpandoc in [this](https://github.com/toroidal-code/yoda/blob/master/src/pandoc/pandoc.ml) so if you want me to add that to the list along with the C# library, I can.

### CODE REVIEW

**Comment by toroidal-code** [comment]
I've also got on ongoing highlighting branch, which adds syntax highlighting bindings for external use. However, it requires Pandoc to export the `Text.Pandoc.Highlighting` module, which it currently isn't doing. If you're interested in integrating that (there's a nice demo in the demo folder), a PR would need to be made changing that one line in `pandoc.cabal`.

**Comment by ShabbyX** [comment]
Thanks a lot! I had seen a message from John saying certain new changes would break other code and I had imagined libpandoc would break. I'm really busy these days though, so I appreciate your help.

By the way, since you seem to know your way around Haskell better than me, can you help me with a debugging issue? The issue is with using non-default settings. Besides being super-sensitive to whitespace (which isn't really a big problem), from the code, it seems that you should be able to give partial settings and it should be able to use default settings for the unspecified ones. However, something with some generic something goes wrong and the end result is that the partial setting gets rejected as invalid and the default settings are used. I tried debugging and/or `trace`ing and just printing the variable that holds the settings results in an exception. Currently, I don't know how to resolve this or what the problem is (the code is relatively simple and seems ok).

Here is a sample code you can use that demonstrates the problem:

```
#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include <pandoc.h>

#define BUF_SIZE 1024

int reader(char *buf, void *user_data)
{
    if (fgets(buf, BUF_SIZE, stdin) == NULL)
        return 0;

    return strlen(buf);
}

void writer(const char *buf, int len, void *user_data)
{
    fwrite(buf, 1, len, stdout);
}

static const char *settings =
"<record name=\"LibPandocSettings\">"
  "<field name=\"writerOptions\">"
    "<record name=\"WriterOptions\">"
      "<field name=\"writerStandalone\">"
        "<int>1</int>"
      "</field>"
    "</record>"
  "</field>"
"</record>";

int main()
{
    pandoc_init();
    char* error = pandoc(BUF_SIZE    /* buffer size */,
            "markdown"    /* input format */,
            "html"        /* output format */,
            settings      /* XML settings */,
            reader        /* the reader function */,
            writer        /* the writer function */,
            NULL          /* private user data */);
    pandoc_exit();

    if (error)
        printf("Error: %s\n", error);
    free(error);

    return 0;
}
```

First of all, do you also see that the writer's standalone option does not take effect? If so, can you figure out what the problem is?

**Comment by toroidal-code** [comment]
It looks like the problem might be in the `Rep` system, Pandoc now has a bunch of types that it just can't handle. Maybes, Colors, etc.
But, Pandoc now derives both  `Generic` and `Typeable` on WriterOptions/ReaderOptions, so a better course of action might be to switch to using something like Text.JSON.Generic, which can handle the full gamut of types transparently, and only needing us to use the `toJSON`/`fromJSON` serializers. I'll see if I can't get a prototype going. 

Let me know what you think. 

Oh, BTW, I left an issue  on @t0yv0's original repo asking them if they would be willing to transfer ownership to you, since forks don't get the issue tracker, and there were a couple I thought of but couldn't make on here.

**Comment by toroidal-code** [comment]
Well, that was _much_ easier than I expected. Here's the [branch](https://github.com/toroidal-code/libpandoc/tree/features/JSON). I'll keep working on for the next week, as I still need to really test it and re-enable XML conversions.

**Comment by ShabbyX** [comment]
Awesome! Thanks a lot. Also, I really can't careless about XML (hard to write for humans, hard to parse for computers), so if you don't have a dependency on it (I don't), I would actually be happy with changing the format to something better. Your call.

Actually t0yv0 offered to do so, but I said things are fine the way they are. Given that I already had a fork with more commits etc, I really didn't know what would happen if ownership is transferred. Would it overwrite my repo? Is it simply a rewrite of the "forked-from" graph? I don't know. In retrospect, it may have not had been a bad idea to try it (after all, I could always restore it with my local replica of the repo).

I just read your issue on t0yv0's repo. I didn't know you cannot open an issue on a fork, so you're right, this is actually a problem! I'll ask him to transfer ownership.
